Tactics/StepThms.lean

```diff
@@ -33,11 +41,190 @@ initialize registerTraceClass `gen_step.print_names
 initialize registerTraceClass `gen_step.debug
 /- When true, prints the number of heartbeats taken per theorem. -/
 initialize registerTraceClass `gen_step.debug.heartBeats
+/- When true, prints the time taken at various steps of generation. -/
+initialize registerTraceClass `gen_step.debug.timing
+
+/-- Assuming that `rawInst` is indeed the right result, construct a proof that
+  `fetch_inst addr state = some rawInst`
+given that `state.program = program` -/
+private def fetchLemma (state program h_program : Expr)
+    (addr : BitVec 64) (rawInst : BitVec 32) : Expr :=
+  let someRawInst := toExpr (some rawInst)
+  mkAppN (mkConst ``fetch_inst_eq_of_prgram_eq_of_map_find) #[
+    state,
+    program,
+    toExpr addr,
+    someRawInst,
+    h_program,
+    mkApp2 (.const ``Eq.refl [1])
+      (mkApp (.const ``Option [0]) <|
+        mkApp (.const ``BitVec []) (toExpr 32))
+      someRawInst
+  ]
+
+-- /-! ## `reduceDecodeInst` -/
+
+/-- `canonicalizeBitVec e` recursively walks over expression `e` to convert any
+occerences of:
+  `BitVec.ofFin w (Fin.mk x _)`
```

Comment:
I feel like I'm missing something. Where could these `BitVec.ofFin ...` terms come from? From `simp/ground` (which I've personally run into)? Anywhere else?

Reply:
`reduce`?

Reply:
It comes from `reduce` indeed, the idea being that instead of painstakingly crafting a simpset that unfold everything but `BitVec.ofNat`, we rather reduce and then refold the `BitVec`s into `BitVec.ofNat`

Reply:
Sounds good --- maybe add this in as a comment.

Reply:
Done, it now reads:
```
/-- `canonicalizeBitVec e` recursively walks over expression `e` to convert any
occurrences of:
  `BitVec.ofFin w (Fin.mk x _)`
to the canonical form:
  `BitVec.ofNat w x` (i.e., `x#w`)

Such expressions tend to result from using `reduce` or
`simp` with `{ground := true}`.
You can call `canonicalizeBitVec` after these functions to ensure you don't
needlessly expose `BitVec` internal details -/
```